File: cart-service/src/main/java/org/vtsukur/graphql/demo/cart/domain/Item.java
package org.vtsukur.graphql.demo.cart.domain;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import javax.persistence.Column;
import javax.persistence.Embeddable;
import java.math.BigDecimal;

@Embeddable
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Item {

    @Column(nullable = false)
    private String productId;

    @Column(nullable = false)
    private int quantity;

    @Column(nullable = false)
    private BigDecimal total;

    Item(String productId, BigDecimal productPrice, int quantity) {
        this.productId = productId;
        this.quantity = quantity;
        this.total = BigDecimal.valueOf(quantity).multiply(productPrice);
    }

}

File: cart-service/src/main/kotlin/org/vtsukur/graphql/demo/cart/domain/Item.kt
package org.vtsukur.graphql.demo.cart.domain

import java.math.BigDecimal
import javax.persistence.Column
import javax.persistence.Embeddable

@Embeddable
data class Item (

    @Column(nullable = false)
    var productId: String = "",

    @Column(nullable = false)
    var quantity: Int = 0,

    @Column(nullable = false)
    var total: BigDecimal = BigDecimal.ZERO) {

    internal constructor(productId: String, productPrice: BigDecimal, quantity: Int) : this() {
        this.productId = productId
        this.quantity = quantity
        this.total = BigDecimal.valueOf(quantity.toLong()).multiply(productPrice)
    }

}
